tasks/run-collectors/run-collectors.yaml

```diff
@@ -0,0 +1,99 @@
+  params:
+    - name: releasePlanAdmissionPath
+      type: string
+      description: Path to the JSON string of the ReleasePlanAdmission in the data workspace
+    - name: releasePlanPath
+      type: string
+      description: Path to the JSON string of the ReleasePlan in the data workspace
+    - name: resultsDir
+      type: string
+      description: The relative path in the workspace to save the collector results to
+  workspaces:
+    - name: data
+      description: Workspace where the CRs are stored
+  steps:
+    - name: run-collectors
+      image: quay.io/konflux-ci/release-service-utils:e633d51cd41d73e4b3310face21bb980af7a662f
+      script: |
+        #!/usr/bin/env bash
+        set -xeo pipefail
+
+        DEFAULT_TIMEOUT=900
+        COLLECTORS_REPOSITORY="https://github.com/SOME_ORG/SOME_REPO" # TODO - update this once repo is created
+        RELEASEPLAN_FILE="$(workspaces.data.path)/$(params.releasePlanPath)"
+        RELEASEPLANADMISSION_FILE="$(workspaces.data.path)/$(params.releasePlanAdmissionPath)"
+        RESULTS_DIR="$(workspaces.data.path)/$(params.resultsDir)"
+        CONCURRENT_LIMIT=4
+
+        execute_collector() { # Expected arguments are [collector json, type of collector]
+            collector="$1"
+            name="$(jq -r '.name' <<< "$collector")"
+            type="$(jq -r '.type' <<< "$collector")"
+            TIMEOUT="$(jq -r --arg default "$DEFAULT_TIMEOUT" '.timeout // $default' <<< "$collector")"
+            ARGS=()
+            NUM_ARGS="$(jq '.params | length' <<< "$collector")"
+            for ((j = 0; j < NUM_ARGS; j++)) ; do
+                param="$(jq -c --argjson j "$j" '.params[$j]' <<< "$collector")"
+                arg_name="$(jq -r '.name' <<< "$param")"
+                arg_value="$(jq -r '.value' <<< "$param")"
+                ARGS=("${ARGS[@]}" --"${arg_name}" "${arg_value}")
+            done
+            # Execute collector
+            timeout "$TIMEOUT" python3 "$type" "${ARGS[@]}" | tee "$RESULTS_DIR"/"$2"-"$name".json
+        }
+
+        if [ ! -d "${RESULTS_DIR}" ] ; then
+            echo The passed results directory does not exist. Failing as there is nowhere to save the results to
+            exit 1
+        fi
+
+        git clone "${COLLECTORS_REPOSITORY}" collectors
+        pushd collectors
+
+        RUNNING_JOBS="\j" # Bash parameter for number of jobs currently running
+
+        if [ ! -f "${RELEASEPLAN_FILE}" ] ; then
+            echo "No valid releasePlan file was found so no tenant collectors will be run."
+        else
+            NUM_TENANT_COLLECTORS="$(jq '.spec.collectors | length' "${RELEASEPLAN_FILE}")"
+            for ((i = 0; i < NUM_TENANT_COLLECTORS; i++)) ; do
+                collector="$(jq -c --argjson i "$i" '.spec.collectors[$i]' "${RELEASEPLAN_FILE}")"
+                # Limit batch size to concurrent limit
+                while (( ${RUNNING_JOBS@P} >= "$CONCURRENT_LIMIT" )); do
+                    wait -n
```

Comment:
Maybe this is not really a concern here, if all of this is reasonable fast.

But one possible improvement would be if instead of `wait -n`, you do `sleep 20s` or something. That way a slow job will not hold up the queue.

Reply:
Can you explain? I thought the way this worked was if there were 4 collectors running with a concurrent limit of 4, it would wait until only 3 were running then start another

Reply:
AFAICS, `wait -n` will wait for the next job in the list, i.e. the first one in the list of currently running jobs. So if the one that is the oldest is slow, then even if the others are fast, we will wait for this one to finish first.

I might be wrong, but this is my understanding from reading the manual.

Reply:
I think it waits for any to finish. Here is my test script
```
#!/usr/bin/env bash

set -exo pipefail

func() {
    echo "Start func call $1 at $(date)"
    sleep $1
    echo "End func call $1 at $(date)"
}

RUNNING_JOBS="\j"

for i in {10,2,3} ; do
    while (( ${RUNNING_JOBS@P} >= 2 )); do
        wait -n
    done
    func $i &
done
```
and here is the output
```
Start func call 10 at Tue Oct  1 08:44:44 PM EDT 2024
Start func call 2 at Tue Oct  1 08:44:44 PM EDT 2024
End func call 2 at Tue Oct  1 08:44:46 PM EDT 2024
Start func call 3 at Tue Oct  1 08:44:46 PM EDT 2024
End func call 3 at Tue Oct  1 08:44:49 PM EDT 2024
End func call 10 at Tue Oct  1 08:44:54 PM EDT 2024
```
func 10 is started first and takes the longest, but func 3 starts as soon as func 2 ends, not waiting for func 10 to finish

Reply:
ok, I stand corrected! Thanks for the test script! It's nice it works like that!